prow/git/git.go

```diff
@@ -0,0 +1,142 @@
+*/
+
+// Package git provides a client to plugins that can do git operations.
+package git
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+)
+
+const github = "https://github.com"
+
+// Client can clone repos. It keeps a local cache, so successive clones of the
+// same repo should be quick. Create with NewClient. Be sure to clean it up.
+type Client struct {
+	// dir is the location of the git cache.
+	dir string
+	// git is the path to the git binary.
+	git string
+	// base is the base path for git clone calls. For users it will be set to
+	// GitHub, but for tests set it to a directory with git repos.
+	base string
+
+	// The mutex protects repoLocks which protect individual repos. This is
+	// necessary because Clone calls for the same repo are racy. Rather than
+	// one lock for all repos, use a lock per repo.
+	// Lock with Client.lockRepo, unlock with Client.unlockRepo.
+	rlm       sync.Mutex
+	repoLocks map[string]*sync.Mutex
+}
+
+// Clean removes the local repo cache. The Client is unusable after calling.
+func (c *Client) Clean() error {
+	return os.RemoveAll(c.dir)
+}
+
+// NewClient returns a client that talks to GitHub. It will fail if git is not
+// in the PATH.
+func NewClient() (*Client, error) {
+	g, err := exec.LookPath("git")
+	if err != nil {
+		return nil, err
+	}
+	t, err := ioutil.TempDir("", "git")
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		dir:       t,
+		git:       g,
+		base:      github,
+		repoLocks: make(map[string]*sync.Mutex),
+	}, nil
+}
+
+func (c *Client) lockRepo(repo string) {
+	c.rlm.Lock()
+	if _, ok := c.repoLocks[repo]; !ok {
+		c.repoLocks[repo] = &sync.Mutex{}
+	}
+	m := c.repoLocks[repo]
+	c.rlm.Unlock()
+	m.Lock()
+}
+
+func (c *Client) unlockRepo(repo string) {
+	c.rlm.Lock()
+	defer c.rlm.Unlock()
+	c.repoLocks[repo].Unlock()
+}
+
+// Clone clones a repository. Pass the full repository name, such as
+// "kubernetes/test-infra" as the repo. This function may take a long time if
+// it is the first time cloning the repo.
+func (c *Client) Clone(repo string) (*Repo, error) {
+	c.lockRepo(repo)
+	defer c.unlockRepo(repo)
+
+	remote := filepath.Join(c.base, repo)
+	cache := filepath.Join(c.dir, repo) + ".git"
+	if _, err := os.Stat(cache); os.IsNotExist(err) {
+		// Cache miss, clone it now.
+		if err := os.Mkdir(filepath.Dir(cache), os.ModePerm); err != nil && !os.IsExist(err) {
+			return nil, err
+		}
+		if b, err := exec.Command(c.git, "clone", "--mirror", remote, cache).CombinedOutput(); err != nil {
+			return nil, fmt.Errorf("git cache clone error: %v. output: %s", err, string(b))
+		}
+	} else if err != nil {
+		return nil, err
+	} else {
+		// Cache hit. Do a git fetch to keep updated.
+		cmd := exec.Command(c.git, "fetch")
+		cmd.Dir = cache
+		if b, err := cmd.CombinedOutput(); err != nil {
+			return nil, fmt.Errorf("git fetch error: %v. output: %s", err, string(b))
+		}
+	}
+	t, err := ioutil.TempDir("", "git")
```

Comment:
The second local clone makes sense but it would be nice to add a comment either here or in the function godoc describing that each Clone invocation, either does a real clone vs a fetch including a local clone from the cache to a separate dir.